File: backend/src/shared/Field.ts
import { FieldType, FieldTypes } from '../core/utils';
import { db } from '@core/db/db';

type Option = {
  value: number;
  text: string;
};

type FieldProps = {
  id?: string;
  name: string;
  type: FieldType;
  relationCollection?: null | string;
  options?: null | Option[];
  isHidden?: boolean;
  isRequired?: boolean;
  isPrimaryKey?: boolean;
  isUnique?: boolean;
};

export default class Field {
  id: string;
  name: string;
  type: FieldType;
  relationCollection: null | string;
  options?: null | Option[];
  isHidden: boolean;
  isRequired: boolean;
  isPrimaryKey: boolean;
  isUnique: boolean;

  constructor({
    id = '',
    name,
    type,
    relationCollection = null,
    options = null,
    isHidden = false,
    isRequired = false,
    isPrimaryKey = false,
    isUnique = false,
  }: FieldProps) {
    this.name = name;
    this.type = type;
    this.relationCollection = relationCollection;
    this.options = options;
    this.isHidden = isHidden;
    this.isRequired = isRequired;
    this.isPrimaryKey = isPrimaryKey;
    this.isUnique = isUnique;

    this.id = id;

    if (id === '') {
      this.id = crypto.randomUUID();
    }
  }

  async exists(collection_id: string) {
    try {
      const result = await db
        .selectFrom('fields_meta as f')
        .select(({ fn }) => fn.countAll().as('length'))
        .where('f.name', '=', this.name)
        .where('f.collection', '=', collection_id)
        .executeTakeFirst();

      if (result) {
        return result.length !== 0;
      }

      return false;
    } catch (e) {
      console.error(e);
    }
  }

  async insertMetaData(collection_id: string) {
    if (!(await this.exists(collection_id))) {
      try {
        await db
          .insertInto('fields_meta')
          .values({
            id: this.id,
            name: this.name,
            type: this.type as string,
            collection: collection_id,
            is_required: this.isRequired,
            relation_collection: this.relationCollection,
            is_hidden: this.isHidden,
            is_primary_key: this.isPrimaryKey,
            is_unique: this.isUnique,
          })
          .execute();

        if (this.type === 'Select' && this.options) {
          for (const option of this.options) {
            await db
              .insertInto('select_options')
              .values({
                id: crypto.randomUUID(),
                collection: collection_id,
                field: this.id,
                text: option.text,
                value: option.value,
              })
              .execute();
          }
        }
      } catch (e) {
        console.error(e);
      }
    }
  }

  sql() {
    const notNull = this.isRequired ? ' NOT NULL' : '';
    const unique = this.isUnique ? ' UNIQUE' : '';

    const col = `"${this.name}" ${FieldTypes[this.type]}${notNull}${unique}`;

    return col.trim();
  }
}

File: Bird/Shared/Field.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using SqlKata.Execution;

public class FieldOption
{
    public int Value { get; set; }
    public string Text { get; set; } = string.Empty;
}

public class FieldProps
{
    public string? Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Type { get; set; } = string.Empty;
    public string? RelationCollection { get; set; }
    public List<FieldOption>? Options { get; set; }
    public bool IsHidden { get; set; } = false;
    public bool IsRequired { get; set; } = false;
    public bool IsPrimaryKey { get; set; } = false;
    public bool IsUnique { get; set; } = false;
}

public class Field
{
    public string Id { get; set; }
    public string Name { get; set; }
    public string Type { get; set; }
    public string? RelationCollection { get; set; }
    public List<FieldOption>? Options { get; set; }
    public bool IsHidden { get; set; }
    public bool IsRequired { get; set; }
    public bool IsPrimaryKey { get; set; }
    public bool IsUnique { get; set; }

    public Field(FieldProps props)
    {
        Name = props.Name;
        Type = props.Type;
        RelationCollection = props.RelationCollection;
        Options = props.Options;
        IsHidden = props.IsHidden;
        IsRequired = props.IsRequired;
        IsPrimaryKey = props.IsPrimaryKey;
        IsUnique = props.IsUnique;

        Id = props.Id ?? IdGenerator.GenerateShortUuid();
    }

    public bool Exists(string collectionId, QueryFactory db)
    {
        try
        {
            var result = db.Query("fields_meta as f")
                .Where("f.name", Name)
                .Where("f.collection", collectionId)
                .Count<int>();

            return result > 0;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Error: {e.Message}");
            return false;
        }
    }

    public void InsertMetaData(string collectionId, QueryFactory db)
    {
        if (!Exists(collectionId, db))
        {
            try
            {
                db.Query("fields_meta")
                    .Insert(new
                    {
                        id = Id,
                        name = Name,
                        type = Type,
                        collection = collectionId,
                        is_required = IsRequired,
                        relation_collection = RelationCollection,
                        is_hidden = IsHidden,
                        is_primary_key = IsPrimaryKey,
                        is_unique = IsUnique
                    });

                if (Type == "Select" && Options != null)
                {
                    foreach (var option in Options)
                    {
                        db.Query("select_options")
                            .Insert(new
                            {
                                id = Guid.NewGuid().ToString(),
                                collection = collectionId,
                                field = Id,
                                text = option.Text,
                                value = option.Value
                            });
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine($"Error: {e.Message}");
            }
        }
    }
}